test/Microsoft.ML.Tests/Scenarios/Api/EstimatorPipe.cs

```diff
@@ -0,0 +1,337 @@
+﻿using Microsoft.ML.Core.Data;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Internal.Utilities;
+using Microsoft.ML.Runtime.Model;
+using Microsoft.ML.Tests.Scenarios.Api;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+[assembly: LoadableClass(typeof(ITransformer), typeof(TransformerChain), null, typeof(SignatureLoadModel),
+    "Transformer chain", TransformerChain.LoaderSignature)]
+
+namespace Microsoft.ML.Tests.Scenarios.Api
+{
+    public sealed class TransformerChain<TLastTransformer> : ITransformer, ICanSaveModel
+        where TLastTransformer : class, ITransformer
+    {
+        private readonly ITransformer[] _transformers;
+        private readonly TransformerScope[] _scopes;
+        public readonly TLastTransformer LastTransformer;
+
+        private const string TransformDirTemplate = "Transform_{0:000}";
+
+        internal TransformerChain(ITransformer[] transformers, TransformerScope[] scopes)
+        {
+            _transformers = transformers.ToArray();
```

Comment:
>.ToArray() [](start = 40, length = 10)

why?

Reply:
we are creating a clone of the input, so that no-one can modify our private fields

---
In reply to: [210996619](https://github.com/dotnet/machinelearning/pull/688#discussion_r210996619) [](ancestors = 210996619)